Client/Frontend/Browser/TabPeekViewController.swift

```diff
@@ -0,0 +1,132 @@
+//
+
+import UIKit
+import Storage
+
+@available(iOS 9.0, *)
+class TabPeekViewController: UIViewController {
+
+    let PreviewActionAddToBookmarks = NSLocalizedString("Add to Bookmarks", comment: "")
+    let PreviewActionAddToReadingList = NSLocalizedString("Add to Reading List", comment: "")
+    let PreviewActionSendToDevice = NSLocalizedString("Send to Device", comment: "")
+    let PreviewActionCopyURL = NSLocalizedString("Copy URL", comment: "")
+    let PreviewActionCloseTab = NSLocalizedString("Close Tab", comment: "")
+
+    let tab: Browser
+    let bvc: BrowserViewController?
+    let tabManager: TabManager
+    var clientPicker: UINavigationController?
+    var isBookmarked: Bool = false
+    var isInReadingList: Bool = false
+    var hasRemoteClients: Bool = false
+
+    // Preview action items.
+    lazy var previewActions: [UIPreviewActionItem] = {
+        var actions = [UIPreviewActionItem]()
+        if !self.isInReadingList {
+            actions.append(UIPreviewAction(title: self.PreviewActionAddToReadingList, style: .Default) { previewAction, viewController in
+                guard let displayURL = self.tab.url?.absoluteString where displayURL.characters.count > 0 else { return }
+                self.bvc?.addToReadingList(displayURL, title: self.tab.lastTitle ?? displayURL)
+            })
+        }
+        if !self.isBookmarked {
+            actions.append(UIPreviewAction(title: self.PreviewActionAddToBookmarks, style: .Default) { previewAction, viewController in
+                guard let displayURL = self.tab.url?.absoluteString where displayURL.characters.count > 0 else { return }
+
+                self.bvc?.addBookmark(displayURL, title: self.tab.lastTitle ?? displayURL)
+                })
+        }
+        if self.hasRemoteClients {
+            actions.append(UIPreviewAction(title: self.PreviewActionSendToDevice, style: .Default) { previewAction, viewController in
+                guard let clientPicker = self.clientPicker else { return }
+                self.bvc?.presentViewController(clientPicker, animated: false, completion: nil)
+                })
+        }
+        actions.append(UIPreviewAction(title: self.PreviewActionCopyURL, style: .Default) { previewAction, viewController in
+            guard let url = self.tab.url where url.absoluteString.characters.count > 0 else { return }
+            let pasteBoard = UIPasteboard.generalPasteboard()
+            pasteBoard.URL = url
+            })
+        actions.append(UIPreviewAction(title: self.PreviewActionCloseTab, style: .Destructive) { previewAction, viewController in
+            guard let tabViewController = viewController as? TabPeekViewController else { return }
+            self.tabManager.removeTab(self.tab)
+            })
+
+        return actions
+    }()
+
+
+    init(tab: Browser, controller: BrowserViewController?, tabManager: TabManager) {
+        self.tab = tab
+        self.bvc = controller
```

Comment:
Passing an entire `BrowserViewController` here will introduce unnecessary coupling. Just use the delegate pattern and have BVC do the glue work?